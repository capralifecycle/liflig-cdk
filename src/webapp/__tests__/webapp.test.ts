import "@aws-cdk/assert/jest"
import { App, Duration, Stack } from "aws-cdk-lib"
import "jest-cdk-snapshot"
import { Webapp } from "../"

test("create webapp with default parameters", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")
  new Webapp(stack, "Webapp", {})
  expect(stack).toMatchCdkSnapshot()
})

test("create webapp with domain and security headers", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")
  new Webapp(stack, "Webapp", {
    domainNames: ["example.com"],
    enableSecurityHeaders: true,
  })
  expect(stack).toMatchCdkSnapshot()
})

test("create webapp with invalid custom security headers", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")
  expect(() => {
    new Webapp(stack, "Webapp", {
      enableSecurityHeaders: true,
      securityHeadersOverrides: {
        contentSecurityPolicy: {
          fontSrc: "\x22",
        },
      },
    })
  }).toThrow()
  expect(() => {
    new Webapp(stack, "Webapp", {
      enableSecurityHeaders: true,
      securityHeadersOverrides: {
        contentSecurityPolicy: {
          fontSrc: ";",
        },
      },
    })
  }).toThrow()
  expect(() => {
    new Webapp(stack, "Webapp", {
      enableSecurityHeaders: true,
      securityHeadersOverrides: {
        contentSecurityPolicy: {
          fontSrc: "\\x22",
        },
      },
    })
  }).toThrow()
})

test("create webapp with domain and response header policy", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")
  new Webapp(stack, "Webapp", {
    domainNames: ["example.com"],
    responseHeadersPolicy: {
      securityHeadersBehavior: {
        strictTransportSecurity: {
          accessControlMaxAge: Duration.days(180),
          override: true,
        },
        contentSecurityPolicy: {
          override: true,
          contentSecurityPolicy: "default-src 'self';",
        },
      },
    },
  })
  expect(stack).toMatchCdkSnapshot()
})
