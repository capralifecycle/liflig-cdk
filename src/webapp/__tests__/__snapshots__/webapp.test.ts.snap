// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create webapp with default parameters 1`] = `
Object {
  "Resources": Object {
    "WebappBucketD38541E4": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "WebappBucketPolicy4224ABFE": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "WebappBucketD38541E4",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "WebappOriginAccessIdentityA216A4CF",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "WebappBucketD38541E4",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "s3:ListBucket",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "WebappOriginAccessIdentityA216A4CF",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "WebappBucketD38541E4",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "WebappDistribution43D777AD": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "CustomErrorResponses": Array [
            Object {
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": Object {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "StackWebappDistributionOrigin1098E590D",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "WebappBucketD38541E4",
                  "RegionalDomainName",
                ],
              },
              "Id": "StackWebappDistributionOrigin1098E590D",
              "OriginPath": "/web",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "WebappOriginAccessIdentityA216A4CF",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "WebappOriginAccessIdentityA216A4CF": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
  },
}
`;

exports[`create webapp with domain and security headers 1`] = `
Object {
  "Resources": Object {
    "WebappBucketD38541E4": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "WebappBucketPolicy4224ABFE": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "WebappBucketD38541E4",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "WebappOriginAccessIdentityA216A4CF",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "WebappBucketD38541E4",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "s3:ListBucket",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "WebappOriginAccessIdentityA216A4CF",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "WebappBucketD38541E4",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "WebappDistribution43D777AD": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            "example.com",
          ],
          "CustomErrorResponses": Array [
            Object {
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": Object {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "FunctionAssociations": Array [
              Object {
                "EventType": "viewer-response",
                "FunctionARN": Object {
                  "Fn::GetAtt": Array [
                    "WebappSecurityHeadersFunctionc85c32e0a9f633a8edb409ebc4166047306b5576786884CD5C",
                    "FunctionARN",
                  ],
                },
              },
            ],
            "TargetOriginId": "StackWebappDistributionOrigin1098E590D",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "WebappBucketD38541E4",
                  "RegionalDomainName",
                ],
              },
              "Id": "StackWebappDistributionOrigin1098E590D",
              "OriginPath": "/web",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "WebappOriginAccessIdentityA216A4CF",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "WebappOriginAccessIdentityA216A4CF": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "WebappSecurityHeadersFunctionc85c32e0a9f633a8edb409ebc4166047306b5576786884CD5C": Object {
      "Properties": Object {
        "AutoPublish": true,
        "FunctionCode": "function handler(event) {
      var response = event.response;
      var headers = response.headers;
      headers['referrer-policy'] = {value: 'strict-origin-when-cross-origin'};
      headers['strict-transport-security'] = {value: 'max-age=63072000;'};
      headers['x-content-type-options'] = {value: 'nosniff'};
      headers['x-frame-options'] = {value: 'DENY'};
      headers['x-xss-protection'] = {value: '1; mode=block'};
      headers['content-security-policy'] = {value: \\"base-uri 'self';child-src 'none';connect-src 'self';default-src 'self';font-src 'self';frame-src 'self';img-src 'self';manifest-src 'self';media-src 'self';object-src 'none';script-src 'self';style-src 'self';\\"};
      return response;
    }",
        "FunctionConfig": Object {
          "Comment": "Functionc85c32e0a9f633a8edb409ebc4166047306b557678",
          "Runtime": "cloudfront-js-1.0",
        },
        "Name": "Functionc85c32e0a9f633a8edb409ebc4166047306b557678",
      },
      "Type": "AWS::CloudFront::Function",
    },
  },
}
`;
